src/cargo/core/shell.rs

```diff
@@ -147,27 +147,66 @@ impl MultiShell {
 }
 
 impl Shell {
-    pub fn create(out: Box<Write + Send>, config: ShellConfig) -> Shell {
+    pub fn create<T: FnMut() -> Box<Write + Send>>(mut out_fn: T, config: ShellConfig) -> Shell {
+        let term = match Shell::get_term(out_fn()) {
+            Ok(t) => t,
+            Err(_) => NoColor(out_fn())
+        };
+
+        Shell {
+            terminal: term,
+            config: config,
+        }
+    }
+
+    /// Create a shell in such a manner that it cannot possibly fail
+    /// Use this after `create` has failed.
+    pub fn create_fallback(out: Box<Write + Send>, config: ShellConfig) -> Shell {
+        Shell {
+            terminal: NoColor(out),
+            config: config,
+        }
+    }
+
+    #[cfg(any(windows))]
+    fn get_term(out: Box<Write + Send>) -> CargoResult<AdequateTerminal> {
+        // Check if the creation of a console will succeed
+        if ::term::WinConsole::new(vec![0u8; 0]).is_ok() {
+            let t = try!(::term::WinConsole::new(out));
+            if !t.supports_color() {
+                Ok(NoColor(Box::new(t)))
+            } else {
+                Ok(Colored(Box::new(t)))
+            }
+        } else {
+            // If we fail to get a windows console, we try to get a `TermInfo` one
+            Shell::get_terminfo_term(out)
+        }
+    }
+
+    #[cfg(any(unix))]
+    fn get_term(out: Box<Write + Send>) -> CargoResult<AdequateTerminal> {
+        Shell::get_terminfo_term(out)
+    }
+
+    fn get_terminfo_term(out: Box<Write + Send>) -> CargoResult<AdequateTerminal> {
```

Comment:
This only returns `Ok`, so it should be fine to remove `CargoResult` from the return type here